File: Official Windows Platform Sample/Windows 8 app samples/[C#]-Windows 8 app samples/C#/Windows 8 app samples/Getting started with C# and Visual Basic sample (Windows 8)/C#/WindowsBlogReader/FeedData.cs
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the Microsoft Public License.
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************


using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Threading.Tasks;
using Windows.Web.Syndication;

namespace WindowsBlogReader
{
    // FeedData
    // Holds info for a single blog feed, including a list of blog posts (FeedItem).
    public class FeedData
    {
        public string Title { get; set; }
        public string Description { get; set; }
        public DateTime PubDate { get; set; }

        private List<FeedItem> _Items = new List<FeedItem>();
        public List<FeedItem> Items
        {
            get
            {
                return this._Items;
            }
        }
    }

    // FeedItem
    // Holds info for a single blog post.
    public class FeedItem
    {
        public string Title { get; set; }
        public string Author { get; set; }
        public string Content { get; set; }
        public DateTime PubDate { get; set; }
        public Uri Link { get; set; }
    }

    // FeedDataSource
    // Holds a collection of blog feeds (FeedData), and contains methods needed to
    // retreive the feeds.
    public class FeedDataSource
    {
        private ObservableCollection<FeedData> _Feeds = new ObservableCollection<FeedData>();
        public ObservableCollection<FeedData> Feeds
        {
            get
            {
                return this._Feeds;
            }
        }

        public async Task GetFeedsAsync()
        {
            Task<FeedData> feed1 =
                GetFeedAsync("http://blogs.windows.com/skydrive/b/skydrive/atom.aspx");
            Task<FeedData> feed2 =
                GetFeedAsync("http://blogs.windows.com/windows/b/windowsexperience/atom.aspx");
            Task<FeedData> feed3 =
                GetFeedAsync("http://blogs.windows.com/windows/b/extremewindows/atom.aspx");
            Task<FeedData> feed4 =
                GetFeedAsync("http://blogs.windows.com/windows/b/business/atom.aspx");
            Task<FeedData> feed5 =
                GetFeedAsync("http://blogs.windows.com/windows/b/bloggingwindows/atom.aspx");
            Task<FeedData> feed6 =
                GetFeedAsync("http://blogs.windows.com/windows/b/windowssecurity/atom.aspx");
            Task<FeedData> feed7 =
                GetFeedAsync("http://blogs.windows.com/windows/b/springboard/atom.aspx");
            Task<FeedData> feed8 =
                GetFeedAsync("http://blogs.windows.com/windows/b/windowshomeserver/atom.aspx");
            // There is no Atom feed for this blog, so use the RSS feed.
            Task<FeedData> feed9 =
                GetFeedAsync("http://blogs.windows.com/windows_live/b/windowslive/rss.aspx");
            Task<FeedData> feed10 =
                GetFeedAsync("http://blogs.windows.com/windows_live/b/developer/atom.aspx");
            Task<FeedData> feed11 =
                GetFeedAsync("http://blogs.windows.com/ie/b/ie/atom.aspx");
            Task<FeedData> feed12 =
                GetFeedAsync("http://blogs.windows.com/windows_phone/b/wpdev/atom.aspx");
            Task<FeedData> feed13 =
                GetFeedAsync("http://blogs.windows.com/windows_phone/b/wmdev/atom.aspx");
            Task<FeedData> feed14 =
                GetFeedAsync("http://blogs.windows.com/windows_phone/b/windowsphone/atom.aspx");

            this.Feeds.Add(await feed1);
            this.Feeds.Add(await feed2);
            this.Feeds.Add(await feed3);
            this.Feeds.Add(await feed4);
            this.Feeds.Add(await feed5);
            this.Feeds.Add(await feed6);
            this.Feeds.Add(await feed7);
            this.Feeds.Add(await feed8);
            this.Feeds.Add(await feed9);
            this.Feeds.Add(await feed10);
            this.Feeds.Add(await feed11);
            this.Feeds.Add(await feed12);
            this.Feeds.Add(await feed13);
            this.Feeds.Add(await feed14);
        }

        private async Task<FeedData> GetFeedAsync(string feedUriString)
        {
            Windows.Web.Syndication.SyndicationClient client = new SyndicationClient();
            Uri feedUri = new Uri(feedUriString);

            try
            {
                SyndicationFeed feed = await client.RetrieveFeedAsync(feedUri);

                // This code is executed after RetrieveFeedAsync returns the SyndicationFeed.
                // Process the feed and copy the data you want into the FeedData and FeedItem classes.
                FeedData feedData = new FeedData();

                if (feed.Title != null && feed.Title.Text != null)
                {
                    feedData.Title = feed.Title.Text;
                }
                if (feed.Subtitle != null && feed.Subtitle.Text != null)
                {
                    feedData.Description = feed.Subtitle.Text;
                }
                if (feed.Items != null && feed.Items.Count > 0)
                {
                    // Use the date of the latest post as the last updated date.
                    feedData.PubDate = feed.Items[0].PublishedDate.DateTime;

                    foreach (SyndicationItem item in feed.Items)
                    {
                        FeedItem feedItem = new FeedItem();
                        if (item.Title != null && item.Title.Text != null)
                        {
                            feedItem.Title = item.Title.Text;
                        }
                        if (item.PublishedDate != null)
                        {
                            feedItem.PubDate = item.PublishedDate.DateTime;
                        }
                        if (item.Authors != null && item.Authors.Count > 0)
                        {
                            feedItem.Author = item.Authors[0].Name.ToString();
                        }
                        // Handle the differences between RSS and Atom feeds.
                        if (feed.SourceFormat == SyndicationFormat.Atom10)
                        {
                            if (item.Content != null && item.Content.Text != null)
                            {
                                feedItem.Content = item.Content.Text;
                            }
                            if (item.Id != null)
                            {
                                feedItem.Link = new Uri(item.Id);
                            }
                        }
                        else if (feed.SourceFormat == SyndicationFormat.Rss20)
                        {
                            if (item.Summary != null && item.Summary.Text != null)
                            {
                                feedItem.Content = item.Summary.Text;
                            }
                            if (item.Links != null && item.Links.Count > 0)
                            {
                                feedItem.Link = item.Links[0].Uri;
                            }
                        }
                        feedData.Items.Add(feedItem);
                    }
                }
                return feedData;
            }
            catch (Exception)
            {
                return null;
            }
        }

        // Returns the feed that has the specified title.
        public static FeedData GetFeed(string title)
        {
            // Simple linear search is acceptable for small data sets
            var _feedDataSource = App.Current.Resources["feedDataSource"] as FeedDataSource;

            var matches = _feedDataSource.Feeds.Where((feed) => feed != null && feed.Title.Equals(title));
            if (matches.Count() == 1) return matches.First();
            return null;
        }

        // Returns the post that has the specified title.
        public static FeedItem GetItem(string uniqueId)
        {
            // Simple linear search is acceptable for small data sets
            var _feedDataSource = App.Current.Resources["feedDataSource"] as FeedDataSource;

            var matches = _feedDataSource.Feeds
                .Where(group => group != null)
                .SelectMany(group => group.Items).Where((item) => item.Title.Equals(uniqueId));
            if (matches.Count() == 1) return matches.First();
            return null;
        }
    }
}

File: Official Windows Platform Sample/Windows 8 app samples/[VB]-Windows 8 app samples/VB/Windows 8 app samples/Getting started with C# and Visual Basic sample (Windows 8)/VB/WindowsBlogReader/FeedData.vb
'*********************************************************
'
' Copyright (c) Microsoft. All rights reserved.
' THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
' ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
' IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
' PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
'
'*********************************************************

Imports Windows.Web.Syndication

' FeedData
' Holds info for a single blog feed, including a list of blog posts (FeedItem).
Public Class FeedData
    Public Property Title() As String = String.Empty
    Public Property Description() As String = String.Empty
    Public Property PubDate() As DateTime

    Private _Items As New List(Of FeedItem)()
    Public ReadOnly Property Items() As List(Of FeedItem)
        Get
            Return Me._Items
        End Get
    End Property
End Class

' FeedItem
' Holds info for a single blog post.
Public Class FeedItem
    Public Property Title() As String = String.Empty
    Public Property Author() As String = String.Empty
    Public Property Content() As String = String.Empty
    Public Property PubDate() As DateTime
    Public Property Link() As Uri
End Class

' FeedDataSource
' Holds a collection of blog feeds (FeedData), and contains methods needed to
' retreive the feeds.
Public Class FeedDataSource
    Private _Feeds As New ObservableCollection(Of FeedData)()
    Public ReadOnly Property Feeds() As ObservableCollection(Of FeedData)
        Get
            Return Me._Feeds
        End Get
    End Property

    Public Async Function GetFeedsAsync() As Task
        Dim feed1 As Task(Of FeedData) =
            GetFeedAsync("http://blogs.windows.com/skydrive/b/skydrive/atom.aspx")
        Dim feed2 As Task(Of FeedData) =
            GetFeedAsync("http://blogs.windows.com/windows/b/windowsexperience/atom.aspx")
        Dim feed3 As Task(Of FeedData) =
            GetFeedAsync("http://blogs.windows.com/windows/b/extremewindows/atom.aspx")
        Dim feed4 As Task(Of FeedData) =
            GetFeedAsync("http://blogs.windows.com/windows/b/business/atom.aspx")
        Dim feed5 As Task(Of FeedData) =
            GetFeedAsync("http://blogs.windows.com/windows/b/bloggingwindows/atom.aspx")
        Dim feed6 As Task(Of FeedData) =
            GetFeedAsync("http://blogs.windows.com/windows/b/windowssecurity/atom.aspx")
        Dim feed7 As Task(Of FeedData) =
            GetFeedAsync("http://blogs.windows.com/windows/b/springboard/atom.aspx")
        Dim feed8 As Task(Of FeedData) =
            GetFeedAsync("http://blogs.windows.com/windows/b/windowshomeserver/atom.aspx")
        ' There is no Atom feed for this blog, so use the RSS feed.
        Dim feed9 As Task(Of FeedData) =
            GetFeedAsync("http://blogs.windows.com/windows_live/b/windowslive/rss.aspx")
        Dim feed10 As Task(Of FeedData) =
            GetFeedAsync("http://blogs.windows.com/windows_live/b/developer/atom.aspx")
        Dim feed11 As Task(Of FeedData) =
            GetFeedAsync("http://blogs.windows.com/ie/b/ie/atom.aspx")
        Dim feed12 As Task(Of FeedData) =
            GetFeedAsync("http://blogs.windows.com/windows_phone/b/wpdev/atom.aspx")
        Dim feed13 As Task(Of FeedData) =
            GetFeedAsync("http://blogs.windows.com/windows_phone/b/wmdev/atom.aspx")
        Dim feed14 As Task(Of FeedData) =
            GetFeedAsync("http://blogs.windows.com/windows_phone/b/windowsphone/atom.aspx")

        Me.Feeds.Add(Await feed1)
        Me.Feeds.Add(Await feed2)
        Me.Feeds.Add(Await feed3)
        Me.Feeds.Add(Await feed4)
        Me.Feeds.Add(Await feed5)
        Me.Feeds.Add(Await feed6)
        Me.Feeds.Add(Await feed7)
        Me.Feeds.Add(Await feed8)
        Me.Feeds.Add(Await feed9)
        Me.Feeds.Add(Await feed10)
        Me.Feeds.Add(Await feed11)
        Me.Feeds.Add(Await feed12)
        Me.Feeds.Add(Await feed13)
        Me.Feeds.Add(Await feed14)
    End Function

    Private Async Function GetFeedAsync(feedUriString As String) As Task(Of FeedData)
        Dim Client As New SyndicationClient
        Dim FeedUri As New Uri(feedUriString)

        Try
            Dim Feed As SyndicationFeed = Await Client.RetrieveFeedAsync(FeedUri)

            ' This code is executed after RetrieveFeedAsync returns the SyndicationFeed.
            ' Process the feed and copy the data you want into the FeedData and FeedItem classes.
            Dim FeedData As New FeedData

            If Feed.Title IsNot Nothing AndAlso Feed.Title.Text IsNot Nothing Then
                FeedData.Title = Feed.Title.Text
            End If
            If Feed.Subtitle IsNot Nothing AndAlso Feed.Subtitle.Text IsNot Nothing Then
                FeedData.Description = Feed.Subtitle.Text
            End If
            If Feed.Items IsNot Nothing AndAlso Feed.Items.Count > 0 Then
                ' Use the date of the latest post as the last updated date.
                FeedData.PubDate = Feed.Items(0).PublishedDate.DateTime
                For Each Item As SyndicationItem In Feed.Items
                    Dim FeedItem As New FeedItem
                    If Item.Title IsNot Nothing AndAlso Item.Title.Text IsNot Nothing Then
                        FeedItem.Title = Item.Title.Text
                    End If
                    FeedItem.PubDate = Item.PublishedDate.DateTime
                    If Item.Authors IsNot Nothing AndAlso Item.Authors.Count > 0 Then
                        FeedItem.Author = Item.Authors(0).Name.ToString()
                    End If
                    ' Handle the differences between RSS and Atom feeds.
                    If Feed.SourceFormat = SyndicationFormat.Atom10 Then
                        If Item.Content IsNot Nothing AndAlso Item.Content.Text IsNot Nothing Then
                            FeedItem.Content = Item.Content.Text
                        End If
                        If Item.Id IsNot Nothing Then
                            FeedItem.Link = New Uri(Item.Id)
                        End If
                    ElseIf Feed.SourceFormat = SyndicationFormat.Rss20 Then
                        If Item.Summary IsNot Nothing AndAlso Item.Summary.Text IsNot Nothing Then
                            FeedItem.Content = Item.Summary.Text
                        End If
                        If Item.Links IsNot Nothing AndAlso Item.Links.Count > 0 Then
                            FeedItem.Link = Item.Links(0).Uri
                        End If
                    End If
                    FeedData.Items.Add(FeedItem)
                Next
            End If
            Return FeedData

        Catch Ex As Exception
            Return Nothing
        End Try
    End Function

    Public Shared Function GetFeed(title As String) As FeedData
        ' Simple linear search is acceptable for small data sets
        Dim _feedDataSource = DirectCast(App.Current.Resources("feedDataSource"), FeedDataSource)

        Dim matches = _feedDataSource.Feeds.Where(Function(feed) feed IsNot Nothing AndAlso feed.Title.Equals(title))
        If matches.Count() = 1 Then
            Return matches.First()
        End If
        Return Nothing
    End Function

    Public Shared Function GetItem(uniqueId As String) As FeedItem
        ' Simple linear search is acceptable for small data sets
        Dim _feedDataSource = DirectCast(App.Current.Resources("feedDataSource"), FeedDataSource)

        Dim matches = _feedDataSource.Feeds.
            Where(Function(group) group IsNot Nothing).
            SelectMany(Function(group) group.Items).Where(Function(item) item.Title.Equals(uniqueId))

        If matches.Count() = 1 Then
            Return matches.First()
        End If
        Return Nothing
    End Function
End Class
